pandas/core/frame.py

```diff
@@ -4922,21 +4922,30 @@ def _arith_op(left, right):
                                  copy=False)
 
     def _combine_match_index(self, other, func, level=None):
+        assert isinstance(other, Series)
         left, right = self.align(other, join='outer', axis=0, level=level,
                                  copy=False)
+        assert left.index.equals(right.index)
         new_data = func(left.values.T, right.values).T
         return self._constructor(new_data,
                                  index=left.index, columns=self.columns,
                                  copy=False)
 
     def _combine_match_columns(self, other, func, level=None, try_cast=True):
+        assert isinstance(other, Series)
         left, right = self.align(other, join='outer', axis=1, level=level,
                                  copy=False)
+        assert left.columns.equals(right.index)
 
-        new_data = left._data.eval(func=func, other=right,
-                                   axes=[left.columns, self.index],
-                                   try_cast=try_cast)
-        return self._constructor(new_data)
+        # Note: we use iloc instead of loc for compat with
+        # case with non-unique columns; same reason why we pin columns
+        # to result below instead of passing it to the constructor.
+        new_data = {n: func(left.iloc[:, n], right.iloc[n])
```

Comment:
> why is moving operations code here a good idea, rather than internals? we want to ove code out of frame, not in

We want to move code out of _internals_.  There are a ton of Issues about `DataFrame` operations that are caused by `Block.eval` not handling things correctly that will be fixed by using the (much more careful and well-tested) `Series` implementations (which BTW are the ones that correctly defer to EA and `FooIndex` implementations).

`DataFrame` ops are all over the place right now: `op(DataFrame, DataFrame)` and `cmp(DataFrame, DataFrame)` operate column-wise (dispatching to `Series` implementation).   Other ops will either a) operate on `self.values` (doing expensive casting for mixed dtypes), or b) dispatch to `_data.eval`, which messes up on many dtypes and weird alignment behavior.